zmon-controller-app/src/main/java/org/zalando/zmon/controller/GrafanaController.java

```diff
@@ -277,14 +278,52 @@ protected static void replaceVariables(JsonNode node, CheckDefinition checkDefin
                 .map(GrafanaController::sanitizeEntityId)
                 .collect(Collectors.joining(","));
 
-        JsonNode node = mapper.readTree(GrafanaController.class.getResourceAsStream("/grafana/dynamic-dashboard.json"));
+        /* allow to create custom dashboards to overwrite generic one */
+        List<GrafanaDashboardSprocService.GrafanaDashboard> dashboards = grafanaService.getGrafanaDashboard("dynamic-zmon-check-" + checkId, "");
+        ObjectNode node;
+
+        if(null != dashboards && dashboards.size() == 1) {
+            node = mapper.createObjectNode();
+            ObjectNode meta = node.putObject("meta");
+            meta.put("type", "db");
+            meta.put("canEdit", true);
+            meta.put("canSave", true);
+            meta.put("canStar", true);
+            meta.put("created", "0001-01-01T00:00:00Z");
+            meta.put("expires", "2999-01-01T00:00:00Z");
+            meta.put("updated", "0001-01-01T00:00:00Z");
+            meta.put("isHome", false);
+            meta.put("slug", id);
+            meta.put("isStarred", false);
+
+            node.set("dashboard", mapper.readTree(dashboards.get(0).dashboard));
+        }
+        else {
+            node = (ObjectNode) mapper.readTree(GrafanaController.class.getResourceAsStream("/grafana/dynamic-dashboard.json"));
+        }
+
         replaceVariables(node, checkDefinition, entityId);
-        ((ObjectNode) node.get("dashboard").get("templating").get("list").get(0)).put("query", entityIds);
-        if (entityId.isPresent()) {
-            // select the right entity in the Grafana templating dropdown
-            final String sanitizedEntityId = sanitizeEntityId(entityId.get());
-            ((ObjectNode) node.get("dashboard").get("templating").get("list").get(0)).putObject("current").put("text", sanitizedEntityId).put("value", sanitizedEntityId);
+
+        if (node.get("dashboard").get("templating").get("list").size() > 0) {
```

Comment:
Extract variable for `node.get("dashboard").get("templating").get("list")` (used immediately afterwards)
Is is safe to descend in that tree like that? Is a missing child element returned as null?